pkg/cloudprovider/aws/capacity.go

```diff
@@ -33,16 +33,20 @@ type Capacity struct {
 	instanceProvider       *InstanceProvider
 	vpcProvider            *VPCProvider
 	launchTemplateProvider *LaunchTemplateProvider
+	instanceTypeProvider   *InstanceTypeProvider
 }
 
 // Create a set of nodes given the constraints.
 func (c *Capacity) Create(ctx context.Context, constraints *cloudprovider.Constraints) ([]cloudprovider.Packing, error) {
-	// 1. Compute Packing given the constraints
-	instancePackings, err := c.packer.Pack(ctx, constraints)
+	// 1. Filter for instance types that fit constraints
+	zoneToInstanceTypes, err := c.instanceTypeProvider.Get(ctx, c.spec.Cluster.Name, constraints)
 	if err != nil {
-		return nil, fmt.Errorf("computing bin packing, %w", err)
+		return nil, fmt.Errorf("filtering instance types by constraints, %w", err)
 	}
 
+	// 2. Compute Packing given the pods and instance types
+	instancePackings := c.packer.Pack(ctx, constraints.Pods, c.instanceTypeProvider.UniqueInstanceTypesFrom(zoneToInstanceTypes))
```

Comment:
I'm a little wary of this flow of calling the instance type provider multiple times and exposing all of the unique/perzone concepts. I need to digest this a little more before I can recommend anything, but I wonder if we can reduce the number of interfaces/concepts we expose from this piece.

Reply:
Likewise to the below comment. Can instanceProvider.Get just return whatever this function needs?

Reply:
Further, w.r.t the accepting NodeCapacities instead of EC2InstanceTypes comment, you'd need to return NodeCapacities. 

Prateek and I have discussed passing real v1.Node objects around (or perhaps simple wrappers w/ a copy function) so that we don't need to keep translating these back and forth. This would further simplify the code on line 76. 

It might be a bigger refactor than what you're up for, but I think it might simplify a lot of the complexity you're forced to contend with here.

Reply:
I did half of this, InstanceTypeProvider now returns a slice of instance types w/ a Zones slice field. The instanceTypes slice is passed directly into the packer and the packer returns it so that Zones is preserved and can be passed directly into the Instance provider for creating the instances.